components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt

```diff
@@ -97,6 +98,54 @@ internal open class TimingDistributionsStorageEngineImplementation(
         }
     }
 
+    /**
+     * Accumulate an array of samples for the provided metric.
+     *
+     * @param metricData the metric information for the timing distribution
+     * @param samples the values to accumulate, provided in the metric's [TimeUnit] (they won't
+     *        be truncated nor converted)
+     * @param timeUnit the [TimeUnit] the samples are in
+     */
+    @Suppress("LongMethod")
+    @Synchronized
+    fun accumulateSamples(
+        metricData: CommonMetricData,
+        samples: LongArray,
+        timeUnit: TimeUnit
+    ) {
+        val validSamples = samples.filter { sample -> sample >= 0 }
+        val numNegativeSamples = samples.size - validSamples.size
+        if (numNegativeSamples > 0) {
+            ErrorRecording.recordError(
+                metricData,
+                ErrorRecording.ErrorType.InvalidValue,
+                "Accumulate $numNegativeSamples negative samples",
+                logger,
+                numNegativeSamples
+            )
+            return
+        }
+
+        // Since the custom combiner closure captures this value, we need to just create a dummy
+        // value here that won't be used by the combine function, and create a fresh
+        // TimingDistributionData for each value that doesn't have an existing current value.
+        val dummy = TimingDistributionData(category = metricData.category, name = metricData.name,
+            timeUnit = timeUnit)
+        validSamples.forEach { sample ->
+            super.recordMetric(metricData, dummy, null) { currentValue, _ ->
+                currentValue?.let {
+                    it.accumulate(sample)
+                    it
+                } ?: let {
+                    val newTD = TimingDistributionData(category = metricData.category, name = metricData.name,
+                        timeUnit = timeUnit)
+                    newTD.accumulate(sample)
+                    return@let newTD
+                }
+            }
```

Comment:
Maybe reduce the code duplication between this and `accumulate` by just calling this from accumulate with an array of length 1?  I'm concerned about these two code paths becoming out-of-sync in the future.

Reply:
Good point. I changed this but I had to leave the negative value check in place. See the comment in the code.